### ISSUE TITLE
We need to be able to customise occurrence upload templates for each biodiversity module

### ISSUE THREAD

**Issue by helendallas** [kartoza/django-bims#3688]
Can we add functionality to make it possible for an administrator to customise the upload template for each biodiversity module. All upload templates would have set of compulsory columns, but then the remainder can be customised as needed.

Each module would have its specific upload template.

**Response from Dimas and Tim:**

At the moment, we already have some columns that are only relevant to algae. With the SANParks project, we will need more of these. And several columns currently used are not relevant.  How can we best deal with this? Is it an option to have an upload template per biodiversity module? 

1) We can add support for unstructured data. Unstructured data are 'attached' to occurrence records, but cannot be used for filtering or displayed in the dashboards etc. 

2) If you need additional fields that can be used  for filtering / display they need to be formally declared in the data model.

**Comment by helendallas** [comment]
@dimasciput The client is needing this urgently.

**Comment by helendallas** [comment]
Here is the revised generic upload template for SANParks

[Occurrence Data Upload Template_SANParks Generic 2024_04_10.xlsx](https://github.com/kartoza/django-bims/files/15234609/Occurrence.Data.Upload.Template_SANParks.Generic.2024_04_10.xlsx)

**Comment by helendallas** [comment]
@dimasciput I am working with Dian to come up with a workable template that will use what we can from ours, and then have some extra columns - some of these will be compulsory, but others optional.  I will attache to this ticket when I have received feedback from her.

**Comment by helendallas** [comment]
@dimasciput An update. I am meeting Di and Judith tomorrow at 09h00. I will then attach the template for SANParks.  

She keeps asking re being able to upload data to a:
-  Park - where there are no Lats and Longs, only the Park or Section Name -  see tickets #3687 and #3844 
-  Where the is no sampling date - see #3844

**Comment by dimasciput** [comment]
Hi @helendallas ,

I need to see the file first to estimate how long it will take to implement this.

- If we have the park uploaded as a layer, then I can probably check the coordinates from the layer matched by name. If there is no name, then just return an error.

- The sampling date is a required field in the system. If we remove this, there will be a lot of complications. I suggest adding a default date, something like 1/1/1970.

**Comment by helendallas** [comment]
The checklist type date will be formatted into the stander SANParks upload template.

It will have most fields completed except no lat/longs, and an artificial sampling date.  SO if you just focus on the SANParks upload template, then the checklist should be covered.

And then the functionality to add unstructured data as discussed above.

**Comment by dimasciput** [comment]
@helendallas 

You can update the template per module here : 

![image](https://github.com/kartoza/django-bims/assets/1979569/b8ccdb65-c093-4101-8695-d5cd070a8526)

Just click the edit taxon group button in the taxon management page. If you leave it empty, then it will use the default template from site setting.

**Comment by helendallas** [comment]
We were not able to test this until the fields have been updated to categorical

**Comment by helendallas** [comment]
Waiting on #3909 to be working. Uploaded the mammals template with extra columns, but upload of test occurrence file not working

**Comment by helendallas** [comment]
None of these extra columns downloaded in the csv. See #3910.

We need all columns in SANParks custom upload template and then the extra ones like conservation status, cites etc.

**Comment by helendallas** [comment]
`this is working, closing ticket as separate ticket for `cites issue. `see #4121
